### PR TITLE
docs: add "Customizing Kubo" doc

### PR DESCRIPTION
**Pull request by guseggert**

### CODE REVIEW

**Comment by BigLep** [comment]
Thanks for starting this Gus!  I think a summary table of pros/cons of the options could be useful too (after have the text in place)

**Comment by BigLep** [comment]
2023-05-05 maintainer conversation: lets also update https://github.com/ipfs/kubo/tree/master/docs/examples/kubo-as-a-library as part of this

**Comment by guseggert** [comment]
> 2023-05-05 maintainer conversation: lets also update https://github.com/ipfs/kubo/tree/master/docs/examples/kubo-as-a-library as part of this

Do you remember what specifically we wanted to update in there?

**Comment by BigLep** [comment]
> > 2023-05-05 maintainer conversation: lets also update https://github.com/ipfs/kubo/tree/master/docs/examples/kubo-as-a-library as part of this
> 
> @BigLep Do you remember what specifically we wanted to update in there?

@guseggert : I don't know exactly.  @aschmahmann may have other thoughts.  I think at the minimum it was to signpost from that document to the new one you have created helping make clear that there are other (and maybe better) ways to proceed than using Kubo as a library.

File: docs/customizing.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+- Using a custom datastore for storing blocks, pins, or other Kubo metadata
+- Adding a custom data transfer protocol into Kubo
+- Customizing Kubo internals, such as adding allowlist/blocklist functionality to Bitswap
+- Adding new commands, interfaces, functionality, etc. to Kubo while reusing the libp2p swarm
+- Building on top of Kubo's configuration and config migration functionality
+
+## Summary
+This table summarizes the tradeoffs between the approaches below:
+
+||  Boxo | Kubo Plugin | Bespoke Extension Point | Go Plugin | Fork |
```
Comment (BigLep) [inline docs/customizing.md:16]: Idea: turn the headers into anchor links to the sections below.

File: docs/customizing.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+* Kubo supports [DNSLink](https://dnslink.dev/) for delegating name->CID mappings to DNS
+
+(This list is not exhaustive.)
+
+These can generally be developed and deployed as sidecars (or full external services) without modifying the Kubo binary.
+
+## Go Plugins
+Go provides [dynamic plugins](https://pkg.go.dev/plugin) which can be loaded at runtime into a Go binary.
+
+Kubo currently works with Go plugins. But using Go plugins requires that you compile the plugin using the exact same version of the Go toolchain with the same configuration (build flags, environment variables, etc.). As a result, generally you'd build Kubo and the plugins together at the same time, and at that point you may as well just compile the functionality directly into Kubo and avoid Go plugins.
```
Comment (BigLep) [inline docs/customizing.md:53]:
```suggestion
Kubo currently works with Go plugins. But using Go plugins requires that you compile the plugin using the exact same version of the Go toolchain with the same configuration (build flags, environment variables, etc.). As a result, you likely need to build Kubo and the plugins together at the same time, and at that point you may as well just compile the functionality directly into Kubo and avoid Go plugins.
```

**Review by BigLep — COMMENTED** [review]
Thanks for writing @guseggert .

How much does the document change when https://github.com/ipfs/kubo/pull/9782 is added?

I leave it to others who know more of the details to review/approve.

**Comment by guseggert** [comment]
> How much does the document change when https://github.com/ipfs/kubo/pull/9782 is added?

We'd add a column to the table and a section describing it

**Review by hacdias — APPROVED** [review]

**Comment by hacdias** [comment]
I am merging this once CI is ✅. Once #9782 is completed, this table should be updated (preferably in the PR itself).

**Pull request merged**